Typing-related tests failure with latest Python 3.5 snapshot

One of the tests fails with the latest Python 3.5 snapshot (what will become 3.5.2):

```
======================================================================
FAIL: test_autodoc.test_type_hints
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/build/sphinx-1.4.1/tests/test_autodoc.py", line 1024, in test_type_hints
    verify_arg_spec(f1, '(x: typing.List[int]) -> typing.List[int]')
  File "/build/sphinx-1.4.1/tests/test_autodoc.py", line 1018, in verify_arg_spec
    eq_(formatargspec(f, *getargspec(f)), expected)
AssertionError: '(x: typing.List[]) -> typing.List[]' != '(x: typing.List[int]) -> typing.List[int]'

----------------------------------------------------------------------
```

Apparently since python/typing@abefbe4a72654e7e385944848b0158969ee33d2e `typing.List[int].__parameters__` is an empty tuple rather than a tuple containing `int`.

On a related note, they changed (in the same commit) the syntax of strings returned by `__repr__` so it is now incompatible with what Sphinx uses (and thus `__repr__` is no longer a sane fallback).

--- a/.travis.yml
+++ b/.travis.yml
@@ -9,6 +9,7 @@ python:
   - "3.3"
   - "3.4"
   - "3.5"
+  - "3.5-dev"
   - "pypy"
 env:
   global:

--- a/sphinx/ext/autodoc.py
+++ b/sphinx/ext/autodoc.py
@@ -269,9 +269,17 @@ def format_annotation(annotation):
         if isinstance(annotation, typing.TypeVar):
             return annotation.__name__
         elif hasattr(typing, 'GenericMeta') and \
-                isinstance(annotation, typing.GenericMeta) and \
-                hasattr(annotation, '__parameters__'):
-            params = annotation.__parameters__
+                isinstance(annotation, typing.GenericMeta):
+            # In Python 3.5.2+, all arguments are stored in __args__,
+            # whereas __parameters__ only contains generic parameters.
+            #
+            # Prior to Python 3.5.2, __args__ is not available, and all
+            # arguments are in __parameters__.
+            params = None
+            if hasattr(annotation, '__args__'):
+                params = annotation.__args__
+            elif hasattr(annotation, '__parameters__'):
+                params = annotation.__parameters__
             if params is not None:
                 param_str = ', '.join(format_annotation(p) for p in params)
                 return '%s[%s]' % (qualified_name, param_str)

